source/server/overload_manager_impl.cc

```diff
@@ -24,16 +25,50 @@ class ThresholdTriggerImpl : public OverloadAction::Trigger {
       : threshold_(config.value()) {}
 
   bool updateValue(double value) override {
-    const bool fired = isFired();
-    value_ = value;
-    return fired != isFired();
+    const OverloadActionState state = actionState();
+    state_.emplace(value >= threshold_ ? OverloadActionState::saturated()
+                                       : OverloadActionState::inactive());
+    return state.value() != actionState().value();
   }
 
-  bool isFired() const override { return value_.has_value() && value_ >= threshold_; }
+  OverloadActionState actionState() const override {
+    return state_.value_or(OverloadActionState::inactive());
+  }
 
 private:
   const double threshold_;
-  absl::optional<double> value_;
+  absl::optional<OverloadActionState> state_;
+};
+
+class RangeTriggerImpl final : public OverloadAction::Trigger {
+public:
+  RangeTriggerImpl(const envoy::config::overload::v3::RangeTrigger& config)
+      : minimum_(config.min_value()), maximum_(config.max_value()) {
+    if (minimum_ >= maximum_) {
+      throw EnvoyException("min_value must be less than max_value");
+    }
+  }
+
+  bool updateValue(double value) override {
+    const OverloadActionState old_state = actionState();
+    if (value <= minimum_) {
+      state_ = OverloadActionState::inactive();
+    } else if (value >= maximum_) {
+      state_ = OverloadActionState::saturated();
+    } else {
+      state_ = OverloadActionState((value - minimum_) / (maximum_ - minimum_));
+    }
+    return state_->value() != old_state.value();
+  }
+
+  OverloadActionState actionState() const override {
+    return state_.value_or(OverloadActionState::inactive());
```

Comment:
maybe just make `state_` non-optional and initialize it as inactive?

Reply:
Great idea, thanks.